envoy/http/async_client.h

```diff
@@ -24,6 +24,28 @@ using FilterConfigSharedPtr = std::shared_ptr<FilterConfig>;
 } // namespace Router
 namespace Http {
 
+class SidestreamWatermarkCallbacks {
```

Comment:
I would prefer to avoid adding the concept of the side streams. It creates a circular dependency - i.e. router depending on async client which in turn depends on the router. It looks very much like the existing decoder/encoder watermark event interface with methods renamed. And it also leaks a concept of side stream into the router filter, which breaks abstraction - i.e. something that is used by ext_proc leaks into the router.

I think you can solve this problem by adding subscription to the encoder watermark events, similar to how it is done for the downstream events.

Reply:
One more thing for consideration:

I took a look at history of  `DecoderFilterWatermarkCallbacks`: It was added in async_client for request mirroring  only  in https://github.com/envoyproxy/envoy/pull/25207/files#diff-e99fc494b853bb5e08e9541feb3b98183f42ecb3a3cdba6a80256f93553ee547

I feel we probably should revert this part https://screenshot.googleplex.com/7YNqz8P5pFna7me and remove `DecoderFilterWatermarkCallbacks` (i.e., replace with new interface here). That interface is limited as it only support push back on downstream because it is implemented as `class StreamDecoderFilterCallbacks : public virtual DecoderFilterWatermarkCallbacks` --- decoder filter only

Reply:
Hi @yanavlasov 

I replaced the `DecoderFilterWatermarkCallbacks` that was introduced  [PR](https://github.com/envoyproxy/envoy/pull/25207/)  with `SidestreamWatermarkCallbacks` and sidestream_watermark_lib

IMHO, it is part of http asyncClient. And from risk management, it's controlled by `setwatermarkcallbacks` and runtime flag. 

However, if you feel strongly about scoping it to ext_proc (which I agree it will be smaller scope of change), I can do that as well.  

PTAL, Thanks